source/server/drain_manager_impl.h

```diff
@@ -29,30 +29,40 @@ class DrainManagerImpl : Logger::Loggable<Logger::Id::main>, public DrainManager
                    Event::Dispatcher& dispatcher);
 
   // Network::DrainDecision
-  bool drainClose() const override;
-  Common::CallbackHandlePtr addOnDrainCloseCb(DrainCloseCb cb) const override;
+  bool drainClose(Network::DrainDirection scope) const override;
+  Common::CallbackHandlePtr addOnDrainCloseCb(Network::DrainDirection direction,
+                                              DrainCloseCb cb) const override;
 
   // Server::DrainManager
-  void startDrainSequence(std::function<void()> drain_complete_cb) override;
-  bool draining() const override { return draining_; }
+  void startDrainSequence(Network::DrainDirection direction,
+                          std::function<void()> drain_complete_cb) override;
+  bool draining(Network::DrainDirection direction) const override {
+    return draining_.load().first && direction <= draining_.load().second;
+  }
   void startParentShutdownSequence() override;
   DrainManagerPtr
   createChildManager(Event::Dispatcher& dispatcher,
                      envoy::config::listener::v3::Listener::DrainType drain_type) override;
   DrainManagerPtr createChildManager(Event::Dispatcher& dispatcher) override;
 
 private:
-  void addDrainCompleteCallback(std::function<void()> cb);
+  void addDrainCompleteCallback(Network::DrainDirection direction, std::function<void()> cb);
 
   Instance& server_;
   Event::Dispatcher& dispatcher_;
   const envoy::config::listener::v3::Listener::DrainType drain_type_;
-
-  std::atomic<bool> draining_{false};
-  Event::TimerPtr drain_tick_timer_;
-  MonotonicTime drain_deadline_;
+  using DrainPair = struct {
+    bool first;
+    Network::DrainDirection second;
+  };
+  std::atomic<DrainPair> draining_{DrainPair{false, Network::DrainDirection::None}};
+  // A map of timers keyed by the direction that triggered the drain
+  std::map<Network::DrainDirection, Event::TimerPtr> drain_tick_timers_;
+  std::map<Network::DrainDirection, MonotonicTime> drain_deadlines_ = {
```

Comment:
This looks like a struct would be a better choice? Are the timer keys dynamic?

Reply:
No we only expect a max of two keys (All and InboundOnly) at any time (and we have an assertion in the code to check).